wagtail/admin/templates/wagtailadmin/generic/chooser/chooser.html

```diff
@@ -65,3 +91,29 @@
         {% endif %}
     </div>
 </div>
+
+
+<script>
```

Comment:
We shouldn't be adding new inline scripts, our reason for using Stimulus is to avoid this. 

We should be able to achieve this kind of logic (or close to it) by changes to the CountController.

Reply:
We can move the entire logic to the frontend `client/src/includes/chooserModal.js`, I just put it and would move it to the frontend after review. so please answer this question to continue work on this PR, can move the logic to the frontend? or make it in the **CountController**, which may change its logic. By the way in all cases it require js to appear and disappear the messages.

Reply:
@elhussienalmasri  I have done some digging and further investigation and feel that the `CountController` is the right home for this as it already does a lot of the work to manage the counts of things.

Firstly, it's best if you take a bit more time to understand what `CountController` is doing and how Stimulus actions work.

The reason the `change` is not working is you had used a data action that only listens for submit - see https://stimulus.hotwired.dev/reference/actions

Updating below (along with other suggested changes further down) should get you close to a point where you can add CSS to hide/show the message blocks instead of adding ad-hoc JS to do this.

```
            {% if is_multiple_choice %}
                <form
                    action="{{ chosen_multiple_url }}"
                    method="GET"
                    data-controller="w-count w-bulk"
                    data-action="change->w-count#count"
                    data-multiple-choice-form
                    data-w-count-active-class="active"
                    data-w-count-find-value="input:checked"
                    data-w-count-max-class="maxiumum"
                    data-w-count-max-value="{{ max_forms }}"
                    data-w-count-offset-value="{{ max_forms_offset }}"
                    data-w-count-over-class="over-maximum"
                >
```

In the controller we can leverage these extra values as follows, supporting an offset and additional classes.

```ts
import { Controller } from '@hotwired/stimulus';
import { ngettext } from '../utils/gettext';

const DEFAULT_ERROR_SELECTOR = '.error-message,.help-critical';

// JSDOC REMOVED FOR NOW
export class CountController extends Controller<HTMLFormElement> {
  static classes = ['active', 'max', 'over'];

  static targets = ['label', 'total'];

  static values = {
    container: { default: '', type: String },
    find: { default: DEFAULT_ERROR_SELECTOR, type: String },
    labels: { default: [], type: Array },
    max: { default: Infinity, type: Number },
    min: { default: 0, type: Number },
    offset: { default: 0, type: Number },
    total: { default: 0, type: Number },
  };

  /** Selector string, used to determine the container/s to search through. If not provided will use the controller's element. */
  declare containerValue: string;
  /** Selector string, used to find the elements to count within the container. */
  declare findValue: string;
  /** Override pluralization strings, e.g. `data-w-count-labels-value='["One item","Many items"]'`. */
  declare labelsValue: string[];
  /** Maximum value, anything equal or over will trigger the 'max' class, anything over will also trigger the 'over' class. */
  declare maxValue: number;
  /** Minimum value, anything equal or below will trigger blank labels in the UI. */
  declare minValue: number;
  /** Offset value to add to the total count when doing calculations */
  declare offsetValue: number;
  /** Total current count of found elements. */
  declare totalValue: number;

  declare readonly activeClasses: string[];
  declare readonly hasLabelTarget: boolean;
  declare readonly hasTotalTarget: boolean;
  declare readonly labelTarget: HTMLElement;
  declare readonly maxClasses: string[];
  declare readonly overClasses: string[];
  declare readonly totalTarget: HTMLElement;

  count() {
    const offset = this.offsetValue;

    const containerSelector = this.containerValue;

    const containers = containerSelector
      ? [...document.querySelectorAll(containerSelector)]
      : [this.element];

    this.totalValue = offset + containers
      .map((element) => element.querySelectorAll(this.findValue).length)
      .reduce((total, subTotal) => total + subTotal, 0);

    return this.totalValue;
  }

  getLabel(total: number) {
    const defaultText = ngettext('%(num)s error', '%(num)s errors', total);

    if (this.labelsValue.length > 1) {
      const [single, plural = this.labelsValue[1], key = '__total__'] =
        this.labelsValue;
      return ngettext(single, plural, total).replace(key, `${total}`);
    }

    return defaultText.replace('%(num)s', `${total}`);
  }

  minValueChanged() {
    this.totalValueChanged(this.count());
  }

  totalValueChanged(total: number) {
    const min = this.minValue;
    const max = this.maxValue;

    if (total > min) {
      this.element.classList.add(...this.activeClasses);
    } else {
      this.element.classList.remove(...this.activeClasses);
    }

    if (total >= max) {
      this.element.classList.add(...this.maxClasses);
      if (total > max) {
        this.element.classList.add(...this.overClasses);
      }
    } else {
      this.element.classList.remove(...this.maxClasses);
      this.element.classList.remove(...this.overClasses);
    }

    if (this.hasLabelTarget) {
      this.labelTarget.textContent = total > min ? this.getLabel(total) : '';
    }
    if (this.hasTotalTarget) {
      this.totalTarget.textContent = total > min ? `${total}` : '';
    }
  }
}

```

In Python it seems like we will need a different value to support this, feel free to pass this from the chooser params directly instead of doing the additional maths in two places.

```py
"max_forms_offset": self.get_max_forms - self.get_max_forms_remainder,
```

This will get you to a point where changing the checkboxes will update classes but it will not fully solve the messaging nuance. I would suggest you get this working first and get to a point where the two (error/help) messages can show & hide.

From here we can opt for 'simpler' messaging to get this over the line or see what other improvements could be made to CountController to support this.

Thank you for keeping at this.

Reply:
here's a commit I used https://github.com/lb-/wagtail/commit/4d3e0a29047e7c18567479c73a34f047e09d566c

Note: the code has an error where it does not correctly remove the classes when dropping back to the maximum, but it hopefully this gives you an idea of what I think could work.

<img width="1787" alt="Screenshot 2024-11-28 at 8 09 32 AM" src="https://github.com/user-attachments/assets/601a6d30-9c7e-4535-b223-2e930c859efe">

Reply:
Further to this, if you set the min value to the offset+ 1. The active class will only trigger when at least one additional item has been selected. 

This way you'll have conditional & dynamic classes for four scenarios. 

1. No selection made (no class added)
2. At least one selection made (active class)
3. Maximum items selected (max class)
4. At least one over the maximum selected (over class)

You'll also have the ability with the labels to have some distinct dynamic labels (see count controller unit tests). And an ability to pipe the total selection number into an element.

I think this can get you super close to what's needed.